lasagne/layers/pool.py

```diff
@@ -601,46 +605,103 @@ class Upscale1DLayer(Layer):
-                self.scale_factor))
+        for i in range(len(self.scale_factor)):
+            if self.scale_factor[i] < 1:
+                raise ValueError('Scale factor must be >= 1, not {0}'.format(
+                    self.scale_factor))
 
         if mode not in {'repeat', 'dilate'}:
             msg = "Mode must be either 'repeat' or 'dilate', not {0}"
             raise ValueError(msg.format(mode))
         self.mode = mode
 
+        if dim > len(self.input_shape):
+            msg = """
+            Invalid input dimension: {0}.
+            Dimension must be less or equal to dimension of input.
+            """
+            raise ValueError(msg.format(dim))
+
+        if dim > 3:
+            msg = """
+            Invalid input dimension: {0}.
+            Number of spatial dimensions greater than 3 are not yet supported.
+            """
+            raise ValueError(msg.format(dim))
+
+        self.dimension = dim
+
     def get_output_shape_for(self, input_shape):
         output_shape = list(input_shape)  # copy / convert to mutable list
-        if output_shape[2] is not None:
-            output_shape[2] *= self.scale_factor[0]
+        output_dim = len(output_shape)
+        ddim = output_dim - self.dimension
+        for i in range(output_dim-1, output_dim - self.dimension - 1, -1):
+            if output_shape[i] is not None:
+                output_shape[i] *= self.scale_factor[i-(ddim)]
         return tuple(output_shape)
 
     def get_output_for(self, input, **kwargs):
-        a, = self.scale_factor
         upscaled = input
         if self.mode == 'repeat':
-            if a > 1:
-                upscaled = T.extra_ops.repeat(upscaled, a, 2)
+            i = len(list(input.shape))-1
+            for a in self.scale_factor:
+                if a > 1:
+                    upscaled = T.extra_ops.repeat(upscaled, a, i)
+                i = i-1
         elif self.mode == 'dilate':
-            if a > 1:
-                output_shape = self.get_output_shape_for(input.shape)
-                upscaled = T.zeros(shape=output_shape, dtype=input.dtype)
-                upscaled = T.set_subtensor(upscaled[:, :, ::a], input)
+            output_shape = self.get_output_shape_for(input.shape)
+            output_dim = len(output_shape)
+            ddim = output_dim - self.dimension
+            slc = [slice(None)] * output_dim
+            for axis in range(output_dim-1, output_dim - self.dimension-1, -1):
+                slc[axis] = slice(0, output_shape[axis],
+                                  self.scale_factor[axis-(ddim)])
+            upscaled = T.zeros(shape=output_shape, dtype=input.dtype)
+            upscaled = T.set_subtensor(upscaled[slc], input)
         return upscaled
 
 
-class Upscale2DLayer(Layer):
+class Upscale1DLayer(UpscaleLayer):
+    """
+    1D upscaling layer
+
+    Calls the class UpscaleLayer with appropriate parameters to perform
+    1D upscaling over the trailing axis of a 3D input tensor
```

Comment:
The docstring should be the same as before. We can point to the base class in a `See also` section (there are examples for the proper docstring syntax in layers/noise.py, for example).